#!/bin/bash
#This script is used to provide a demo for perfroming 
#	dds_analysis dTarget_methy_vs_express
#	on rat data by integrating DMR and DEG data together
#	and predicting putative target genes to a DMR based on their 
#	associations from either TSS or 5'distance reginos.

#main path of input data
IN_DATA_PATH='../../data/hap1_cell/in_data/final_demo_data/hap1_cell/'

#path of DMR results from dmr_analysis
IN_MR_PATH=${IN_DATA_PATH}'/out_data/DMR_CpG_context/'

#path of DEG results from bpb3
IN_DEG_PATH=${IN_DATA_PATH}'/in_data/DEG/'

#DEG file name from bpb3 differential_analysis, the original DEF file from bpb3 that was used to convert Zscores in dds_analysis preprocess 
IN_DEG_FILE='HAP1_P1_vs_HAP1_KO1_differentially_expressed_genes_min0Fd_min0RPKM.txt '
in_data_str='_hap1'

#path to output data
OUT_PATH='../../data/hap1_cell/out_data/'

#path to exported MRs that are not located in TSS or enhancer regions
FILE_FOLD=${OUT_PATH}/out4mr_not_in_tss_enhancer
#file name for background sample list that contain all MRs not located in TSS or enhancers
BACK_FILE=${OUT_PATH}/background_samples_list.tsv

#whether to skip below two steps in the pipeline
is_run_dmr_export=1 # 1 for exporting, other values for skipping this step 
is_run_dtarget=1    # 1 for run dTarget prediction , other values for skipping this step

if [ $is_run_dmr_export == 1 ];
then
##15 here two input file are generated by preprocess_data4dds_test.py 
#1. export data of dmr regions that are located in either TSS or 5distance regions by using dmr_analysis
dmr_analysis dmr_exportData \
        --input_mr_data_folder ${IN_MR_PATH} \
        --output_file_folder ${OUT_PATH}/out4dmr_in_deg_tss_5dist \
        --input_file_format 0 \
        --number_of_processes 10 --input_file ${OUT_PATH}/uqdmr_regions_in_deg_tss_5dist${in_data_str}.bed -wtStr 'HAP1_P_'
echo "Export data of DMRs overlapping to TSS or 5distnace - Done "
echo ""

#2. export data of mrs that are not in tss and enhancer 
dmr_analysis dmr_exportData  \
        --input_mr_data_folder ${IN_MR_PATH} \
        --output_file_folder ${OUT_PATH}/out4mr_not_in_tss_enhancer \
        --input_file_format 0 \
        --number_of_processes 10 --input_file ${OUT_PATH}/mr_regions_not_in_enhancers${in_data_str}_tss.bed -wtStr 'HAP1_P_'
echo "Export data of MRs not in TSS or enhancers - Done "
echo ""

fi
#end export data

#16. 
#3. create background file list if it is not exist.
#for large list of files, bash script may not work but a python script is needed!
if ! [ -f $BACK_FILE ];
then
  echo $BACK_FILE " not exists and create one ! "
  if [ -e $FILE_FOLD ];
  then
    	ls  ./${FILE_FOLD}/chr*/data/*raw*.* > $BACK_FILE
        echo "Create background file list in " $BACK_FILE 
  else
    	echo "Cannot create background file because no data folder find! " $FILE_FOLD
  fi
fi

#17. 
#4. to run dds_analysis dTarget_methy_vs_express for predicting putative target genes to DMRs based on gene expression profiles
# unique gene to DMR file prepared by  preprocess_data4dds_test.py
gene_mr_file=${OUT_PATH}/uqGeneDmr_regions_in_deg_tss${in_data_str}.bed

#prepare a tab delimiated gene expression file in which the group mean values and rratio are added.
#this file will be used to plot average methylation levels of selected gene in TSS and Enhancer regions
#After inputting a DEG file  exported by bpb3 differential_expression, it exports a tab delimiated file by adding three columns values of the group mean and rratio.
#This filtered DEG file will only be used in plot_tss_enhancer_mrs for exporting data
#this function only consider input data as RPKM values
dds_analysis filterDEG4bpb3 --in_group1_str 'HAP1_P1' --in_group2_str 'HAP1_KO1' \
        --in_folder ${IN_DEG_PATH} \
        --in_file ${IN_DEG_FILE} \
        --min_median_RPKM 0 --rr_cutoff 0.0

#we can skip this manual input step if know the input gene expression file name 
if [ 1 == 2 ];
then
echo ""
read -p "To continue please copy the exported zscore cluster file name and path from bpb3 filterDEG4bpb3 then click return: " gene_exp_file
echo ""
echo ""
echo "gene_exp_file is :  $gene_exp_file "
echo ""
read -p "To continue please copy the exported group mean file name and path from bpb3 filterDEG4bpb3 then click return: " IN_DEG_FILE
echo ""
echo ""
echo "IN_DEG_FILE is :  $IN_DEG_FILE "
echo ""
echo ""
fi
#end test

#we assume know the input gene exp file name
gene_exp_file0=${IN_DEG_PATH}/${IN_DEG_FILE}
#here we assume file name end with .txt
finds='.txt'
replace1='_rratio_filtered4cluster.csv'
replace2='_rratio_filtered.csv'
gene_exp_file=${gene_exp_file0//$finds/$replace1}
IN_DEG_FILE=${gene_exp_file0//$finds/$replace2}
echo ""
echo "gene_exp_file is :  $gene_exp_file "
echo ""
echo "IN_DEG_FILE is :  $IN_DEG_FILE "
echo ""

# path of DMRs associated with DEG, TSS and 5distance, prepared by run dds_analysis preprocess
in_mr_data_folder=${OUT_PATH}/out4dmr_in_deg_tss_5dist
# a file for a list of background samples
in_background_mr_file=$BACK_FILE
#number of random sampling for the test
number_of_samples=10

if [ $is_run_dtarget == 1 ];
then

#18. here for test purpose pval reg_cutoff >0.05 , but it shall use a pval at least <0.05 and number of samples >=1000 in real prediction ! 
#5. to test target gene and DMRs assocation from TSS regions
dds_analysis dTarget_methy_vs_express -inGeneMRfile $gene_mr_file  -mrTAB  \
	-inGeneEXPfile $gene_exp_file -expTAB \
	-inMRfolder $in_mr_data_folder -outName 'tss_region_' \
	-output_path $OUT_PATH -sampleName sample_name4replace.tsv \
	-pathDepth 1 -inBackgroundList $in_background_mr_file -reg_cutoff 0.5 -cutoff 1.0 -totalSamples $number_of_samples -numOfprocesses 10

echo "Done with TSS target gene prediction"
echo ""

#19. 
#6. to test target gene and DMRs associatioin from  5dist regions
gene_mr_file=${OUT_PATH}/uqGeneDmr_regions_in_deg_5dist${in_data_str}_overlap_enhancer.bed
dds_analysis dTarget_methy_vs_express -inGeneMRfile $gene_mr_file -mrTAB  \
	-inGeneEXPfile $gene_exp_file -expTAB \
	-inMRfolder $in_mr_data_folder -outName 'distance_region_'  \
	 -output_path $OUT_PATH -sampleName sample_name4replace.tsv \
	-pathDepth 1 -inBackgroundList $in_background_mr_file -reg_cutoff 0.5 -cutoff 1.0 -totalSamples $number_of_samples -numOfprocesses 10

echo "Done with 5distance region target gene prediction" 
echo ""

fi
#end dTarget run


#20. 
#7. plot selected target gene and DMR associatoins
echo ""
echo ${gene_exp_file}
echo  ${OUT_PATH}/out4dmr_in_deg_tss_5dist 
dds_analysis plot_mr_vs_exp -inGeneEXPfile ${gene_exp_file}  \
        -dpi 300 -inMRfolder ${OUT_PATH}/out4dmr_in_deg_tss_5dist \
	-expTAB -inGene 'TRIM32' -inMR 'chr9:mr104' -wtStr 'HAP1_P' -output_path ${OUT_PATH}
echo "Done with plot_vs_exp "
echo ""

#8. plot avearge methylation in TSS and enhancer regions for selected target gene
dds_analysis plot_tss_enhancer_mrs \
	-exp_file $IN_DEG_FILE \
	-dmr_file ${IN_MR_PATH}/3_chroms_all_mr_data_range_dmrRanking.bed  \
	-tss_file ${OUT_PATH}/tss_region_10sampling.csv  \
	-enc_file ${OUT_PATH}/distance_region_10sampling.csv \
	-is_negative 1 -genes 'MRPS25,PLCL2,TRIM32' -mr_folder ${OUT_PATH}/out4dmr_in_deg_tss_5dist/ \
	-folder_name '' --dmr_file_not_compressed \
	-gX 2000 -gY 1000 -wtStr 'HAP1_P_' \
	-out_folder ${OUT_PATH}/plot_tss_enhancer_mrs
echo "Done with plot_tss_enhancer_mrs"


