<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>plot_tss_enhancer_mrs &mdash; dds_analysis 1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="filterDEG4bpb3" href="filterDEG4bpb3.html" />
    <link rel="prev" title="plot_mr_vs_exp" href="plot_mr_vs_exp.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            dds_analysis
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="bpb3summary2bed_format.html">bpb3summary2bed_format</a></li>
<li class="toctree-l1"><a class="reference internal" href="map_block2genome.html">map_block2genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="map_block2chromSegment.html">map_block2chromSegment</a></li>
<li class="toctree-l1"><a class="reference internal" href="map_block2dmr.html">map_block2dmr</a></li>
<li class="toctree-l1"><a class="reference internal" href="find_geneExp4block.html">find_geneExp4block</a></li>
<li class="toctree-l1"><a class="reference internal" href="find_block_patientID.html">find_block_patientID</a></li>
<li class="toctree-l1"><a class="reference internal" href="combine_dmr_deg2block.html">combine_dmr_deg2block</a></li>
<li class="toctree-l1"><a class="reference internal" href="filter_blocks.html">filter_blocks</a></li>
<li class="toctree-l1"><a class="reference internal" href="collect_gene_names4blocks.html">collect_gene_names4blocks</a></li>
<li class="toctree-l1"><a class="reference internal" href="check_block_gene_inTAD.html">check_block_gene_inTAD</a></li>
<li class="toctree-l1"><a class="reference internal" href="dds_geneRanking.html">dds_geneRanking</a></li>
<li class="toctree-l1"><a class="reference internal" href="go_pathway_analysis4out_blocks_gene.html">go_pathway_analysis4out_blocks_gene</a></li>
<li class="toctree-l1"><a class="reference internal" href="find_enhancer_target_genes.html">find_enhancer_target_genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="chromSegment_test4blocks.html">chromSegment_test4blocks</a></li>
<li class="toctree-l1"><a class="reference internal" href="dTarget_methy_vs_express.html">dTarget_methy_vs_express</a></li>
<li class="toctree-l1"><a class="reference internal" href="plot_mr_vs_exp.html">plot_mr_vs_exp</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">plot_tss_enhancer_mrs</a></li>
<li class="toctree-l1"><a class="reference internal" href="filterDEG4bpb3.html">filterDEG4bpb3</a></li>
<li class="toctree-l1"><a class="reference internal" href="preprocess.html">preprocess</a></li>
<li class="toctree-l1"><a class="reference internal" href="demoFL.html">demoFL</a></li>
<li class="toctree-l1"><a class="reference internal" href="demoHAP1.html">demoHAP1</a></li>
<li class="toctree-l1"><a class="reference internal" href="demoRAT.html">demoRAT</a></li>
<li class="toctree-l1"><a class="reference internal" href="demoFLMR.html">demoFLMR</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">dds_analysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">plot_tss_enhancer_mrs</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/content/plot_tss_enhancer_mrs.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="plot-tss-enhancer-mrs">
<h1>plot_tss_enhancer_mrs<a class="headerlink" href="#plot-tss-enhancer-mrs" title="Permalink to this heading"></a></h1>
<p>This module plots the average DMR methylation levels at TSS and enhancer regions based on their predicted target genes from dTarget_methy_vs_express. It also exports differential gene and differential methylation information for each pair of putative targets and TSS or enhancers.</p>
<p>Optional arguments:</p>
<ul class="simple">
<li><p><cite>-h, –help</cite>: Show the help message and exit.</p></li>
</ul>
<p>Required arguments:</p>
<ul class="simple">
<li><p><cite>-exp_file DIFFERENTIAL_GENE_EXPRESSION_FILE, –differential_gene_expression_file</cite>: A tab-delimited file with at least four columns (e.g., gene, group1_mean, group2_mean, rratio/log2folder change) containing differential gene expression information, which is exported by (bpb3 filterDEG4bpb3). If more than four columns are included in this file, the program will assume the first column is gene, and the last 3 columns are group1_mean, group2_mean, and rratio/log2folder change, respectively.</p></li>
<li><p><cite>-genes SELECTED_GENES4PLOT, –selected_genes4plot SELECTED_GENES4PLOT</cite>: A string of gene symbols that will be selected for plotting the average methylation levels in their predicted regulatory regions such as TSS and Enhancer. Gene symbols are separated by commas (,) such as gene1,gene2,gene3.</p></li>
</ul>
<p>Optional arguments:</p>
<ul class="simple">
<li><p><cite>-dmr_file DIFFERENTIAL_METHYLATION_FILE, –differential_methylation_file</cite>: A tab-delimited file with only five columns (e.g., chrom, start_pos, end_pos, infos, pval) containing differential methylation regions, which is exported by (dmr_analysis dmr_combine_multChrs4rank). For the column “infos,” the data structure is chrom:mr_id:anything:hyper/hypo:anything, where the information in the first, second, and fourth strings separated by “:” will be used by the program for finding DMRs.</p></li>
<li><p><cite>-tss_file TSS_TARGET_FILE, –tss_target_file TSS_TARGET_FILE</cite>: A tab-delimited file with nine columns (index, gene, mr_id, rg_Tvalue, rg_Pvalue, enhancer, gene_type, utest_Pvalue, expectPvalue) containing predicted information of TSS target genes by using (dds_analysis dTarget_methy_vs_express).</p></li>
<li><p><cite>-enc_file ENHANCER_TARGET_FILE, –enhancer_target_file ENHANCER_TARGET_FILE</cite>: A tab-delimited file with nine columns (index, gene, mr_id, rg_Tvalue, rg_Pvalue, enhancer, gene_type, utest_Pvalue, expectPvalue) containing predicted information of enhancer target genes by using (dds_analysis dTarget_methy_vs_express).</p></li>
<li><p><cite>-mr_folder MR_DATA_FOLDER, –mr_data_folder MR_DATA_FOLDER</cite>: A file folder containing all DNA methylation data of methylated regions exported by dmr_analysis dmr_exportData.</p></li>
<li><p><cite>-folder_name PREFIX_FOLDERNAME4CHROM, –prefix_foldername4chrom PREFIX_FOLDERNAME4CHROM</cite>: A prefix folder name for exported methylation data of each chromosome under -mr_folder (e.g., prefix_foldername4chrom + chr1 means all methylation data for chr1).</p></li>
<li><p><cite>-out_folder, –out_data_folder</cite>: Output folder name for exported data from the current analysis. If this folder does not exist in the input data folder, it will be created. Default is “data”.</p></li>
<li><p><cite>-out_plots, –out_plot_folder</cite>: Output folder name for all plots. Default is “plots”.</p></li>
<li><p><cite>-is_negative, –is_negative_target</cite>: Select plot for negative, positive, or both responses of predicted DMR targets. Use 0, 1, or 2</p></li>
</ul>
<p>for positive, negative, and all responses, respectively. Default is 1 for negative targets.</p>
<ul class="simple">
<li><p><cite>-gX, –gene_upstream_X</cite>: Set the length of upstream to transcription start side (TSS) in the plot. Default is 500.</p></li>
<li><p><cite>-gY, –gene_downstream_Y</cite>: Set the length of downstream to transcription start side (TSS) in the plot. Default is 500.</p></li>
<li><p><cite>-G, –gene_or_enhancer_length</cite>: Set the gene or enhancer length in the plot. Default is 2000.</p></li>
<li><p><cite>-w, –step_size_in_region</cite>: Step size in a region for data smoothing. Default is 100.</p></li>
<li><p><cite>-window, –window_size_in_combined_regions</cite>: Window size of combined regions such as TSS and Enhancer. Default is 600. Increase this parameter if there is a gap between two combined regions (&gt;500).</p></li>
<li><p><cite>-sigma, –data_smooth_sigma</cite>: Data smooth parameter sigma. Default is 50.</p></li>
<li><p><cite>-flank, –flank_region</cite>: Flank region to add on the two sides of a promoter or enhancer region. Default is 100.</p></li>
<li><p><cite>-MRs_txt, –MRs_txt_description</cite>: Text description of MRs. Default is “DMRs”.</p></li>
<li><p><cite>-dmr_compressed, –dmr_file_not_compressed</cite>: Specify this option if the DMR file is not compressed. Default is False (dmr_file is compressed).</p></li>
<li><p><cite>-flip_rr, –not_flip_sign_of_rratio</cite>: Specify this option if you do not want to flip the sign of rratio. Default is False (sign of rratio is flipped).</p></li>
</ul>
<p>Please replace the angle-bracketed placeholders (<cite>&lt;&gt;</cite>) with appropriate values when using the command.</p>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="plot_mr_vs_exp.html" class="btn btn-neutral float-left" title="plot_mr_vs_exp" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="filterDEG4bpb3.html" class="btn btn-neutral float-right" title="filterDEG4bpb3" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Junbai and Omer.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>